src/drivers/intel/cavs/ipc.c

```diff
@@ -133,12 +143,56 @@ static void ipc_irq_handler(void *arg)
 	}
 }
 
+#if CAVS_VERSION >= CAVS_VERSION_1_8
+static struct sof_ipc_cmd_hdr *ipc_cavs_read_set_d0ix(uint32_t dr, uint32_t dd)
+{
+	struct sof_ipc_pm_pwr_gating *cmd = _ipc->comp_data;
+
+	cmd->hdr.cmd = SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_PWR_GATING;
+	cmd->hdr.size = sizeof(*cmd);
+	cmd->flags = dd & CAVS_IPC_MOD_SETD0IX_BIT_MASK;
+
+	return &cmd->hdr;
+}
+
+static struct sof_ipc_cmd_hdr *ipc_cavs_read_msg(void)
+{
+	struct sof_ipc_cmd_hdr *hdr = NULL;
+	uint32_t dr;
+	uint32_t dd;
+
+	dr = ipc_read(IPC_DIPCTDR);
+	dd = ipc_read(IPC_DIPCTDD);
+
+	/* if there is no cAVS module IPC in regs go the previous path */
+	if (!(dr & CAVS_IPC_MSG_TGT))
+		return mailbox_validate();
+
+	switch (CAVS_IPC_TYPE_S(dr)) {
+	case CAVS_IPC_MOD_SET_D0IX:
+		hdr = ipc_cavs_read_set_d0ix(dr, dd);
```

Comment:
It is OK to reply via mailbox since the DSP may guarantee the underlying memory is on. There are constraints, described in the commit message, that force us to reply via mailbox and keep that on atm until the cAVS IPC driver is modified to use both pairs of registers.